In [1]:
import os
try:
    os.chdir('src')
except:
    pass
print(os.getcwd())

/home/shamvinc/ssl_time_series/mvts_transformer/src


In [2]:
%load_ext autoreload
%autoreload 2
%reload_ext autoreload


<center><img src="../img/img_0.PNG"  width="1000" height="240"/></center>


# SimMTM is a simple self-supervised learning framework for Masked Time-Series Modeling.

In [3]:
import torch
import torch.nn as nn
import torch.nn.functional as F

import numpy as np
from tqdm import tqdm
import copy



from datasets.datasplit import split_dataset
from datasets.data import data_factory, Normalizer, TSRegressionArchive, CSVRegressionArchive
from datasets.datasplit import split_dataset
from datasets.dataset import collate_superv
from models.ts_transformer import model_factory
from models.loss import get_loss_module, contrastive_loss
from optimizers import get_optimizer

from options import Options
from running import setup


# Masked Modeling

<img src="../img/img_1.PNG"  width="900" height="240"/>

<img src="../img/img_2.PNG"  width="900" height="240"/>

### Random Masking is not a good choice to learn a good representation. 

<img src="../img/img_3.PNG" width="600"/>

### Therefore, we randomly mask a sequence of points in a serie.

In [4]:
def geom_noise_mask_single(L, lm, masking_ratio):
    """
    Randomly create a boolean mask of length `L`, consisting of subsequences of average length lm, masking with 0s a `masking_ratio`
    proportion of the sequence L. The length of masking subsequences and intervals follow a geometric distribution.
    Args:
        L: length of mask and sequence to be masked
        lm: average length of masking subsequences (streaks of 0s)
        masking_ratio: proportion of L to be masked

    Returns:
        (L,) boolean numpy array intended to mask ('drop') with 0s a sequence of length L
    """
    keep_mask = np.ones(L, dtype=bool)
    p_m = 1 / lm  # probability of each masking sequence stopping. parameter of geometric distribution.
    p_u = p_m * masking_ratio / (1 - masking_ratio)  # probability of each unmasked sequence stopping. parameter of geometric distribution.
    p = [p_m, p_u]

    # Start in state 0 with masking_ratio probability
    state = int(np.random.rand() > masking_ratio)  # state 0 means masking, 1 means not masking
    for i in range(L):
        keep_mask[i] = state  # here it happens that state and masking value corresponding to state are identical
        if np.random.rand() < p[state]:
            state = 1 - state

    return keep_mask

# SimMTM ultilizes both contrastive learning and mask modeling to learn the data representation.


# 1 - Contrastive learning

<img src="../img/img_5.png"/>

### The contrastive loss is the following: (Eq. 8 in the paper)

<center><img src="../img/img_6.PNG"/><center/>

In [5]:
def demo_contrastive_loss(z, batch_size):
    s = s.squeeze(-1) 

    B = s.shape[0]
    v = s.reshape(B, -1)

    norm_v = torch.norm(v, p=2, dim=-1).unsqueeze(-1)
    v = v/norm_v
    u = torch.transpose(v, 0, 1)

    R = torch.matmul(v,u)

 
    R = torch.exp(R/tau) # (batch + mask size) x (batch + mask size)
    
    # number of masks
    M = B//batch_size
    mask = torch.eye(batch_size, device=R.device).repeat_interleave(M,dim=0).repeat_interleave(M,dim=1)

    denom = R * (torch.ones_like(R) - torch.eye(R.shape[0], device=R.device))

    denom = R.sum(-1).unsqueeze(-1)

    loss = torch.log(R/denom)
    

    loss = (loss * (mask - torch.eye(R.shape[0], device=R.device))).sum(1)/(M-1) # except no masked unit
    loss = loss.mean(0)
    
    return -loss


# 2 - Masked Modeling

### SimMTM proposes to recover a time serie by the weighted sum of multiple masked points, which eases the reconstruction task by assembling ruined but complementary temporal variations.

<img src="../img/img_4.png"/>

In [6]:


class DemoSimMTMTransformerEncoder(nn.Module):

    def forward(self, X):
        """
        Args:
            X: (batch_size, seq_length, feat_dim) torch tensor of masked features (input)
            padding_masks: (batch_size, seq_length) boolean tensor, 1 means keep vector at this position, 0 means padding
        Returns:
            output: (batch_size, seq_length, feat_dim)
        """

        # permute because pytorch convention for transformers is [seq_length, batch_size, feat_dim]. padding_masks [batch_size, feat_dim]
        _x = X
     
        for i in range(self.temporal_unit):
            mask = geom_noise_mask_single(X.shape[0] * X.shape[1] * X.shape[2], 5, 0.3)
            mask = mask.reshape(X.shape[0], X.shape[1], X.shape[2])
            mask = torch.from_numpy(mask).to(X.device)
            x_masked = mask * X
            _x = torch.cat([_x, x_masked], axis=-1) # [batch_size, seq_length, feat_dim * temporal_unit]
    
        
        _x = _x.reshape(X.shape[0] * (self.temporal_unit + 1), X.shape[1], X.shape[2])
  

        inp = _x.permute(1, 0, 2)
        inp = self.project_inp(inp) * math.sqrt(self.d_model)  # [seq_length, batch_size, d_model] project input vectors to d_model dimensional space
        inp = self.pos_enc(inp)  # add positional encoding

        
        output = self.transformer_encoder(inp)  # (seq_length, batch_size, d_model)
        output = self.act(output)  # the output transformer encoder/decoder embeddings don't include non-linearity
        output = output.permute(1, 0, 2)  # (batch_size, seq_length, d_model)
        output = self.dropout1(output)

        z_hat = self.project(output)
        # Most probably defining a Linear(d_model,feat_dim) vectorizes the operation over (seq_length, batch_size).
        output = self.output_layer(z_hat)  # (batch_size, seq_length, feat_dim)

        return output

    
    def project(self, z, tau=0.02):
        _z = z.transpose(1, 2) # [batch_size, d_model, seq_length]
        s = self.projector_layer(_z) # [batch_size, d_model, 1]
        s = s.squeeze(-1) 

        B = s.shape[0]
        v = s.reshape(B, -1)

        norm_v = torch.norm(v, p=2, dim=-1).unsqueeze(-1)
        v = v/norm_v
        u = torch.transpose(v, 0, 1)
        
        R = torch.matmul(v,u)
     
  
        R = torch.exp(R/tau) # (batch + mask size) x (batch + mask size)
        R = R * (torch.ones_like(R) - torch.eye(R.shape[0], device=R.device)) # zero out the weight of no masked component
        R = R/R.sum(-1).unsqueeze(-1)
        M = self.temporal_unit + 1
        R = R[::M] # extract every no mask unit # (batch size) x (batch + mask size)

        z_hat = (R.unsqueeze(-1).unsqueeze(-1) * z.unsqueeze(0)).sum(1) 
        return z_hat




# Data Loading and Preparation

In [7]:
args = Options().parse()  # `argsparse` object
# args.data_dir = '../datasets/BeijingPM25Quality'
args.data_dir = '../datasets/AppliancesEnergy'
 
args.task = 'regression'
args.output_dir = '../experiments'
config = setup(args)
# config = setup(args)  # configuration dictionary

2023-08-21 00:55:10,648 | INFO : Stored configuration file in '../experiments/_2023-08-21_00-55-10_Btc'


In [8]:
data = TSRegressionArchive(config['data_dir'], pattern='TRAIN', config=config)
test_data = TSRegressionArchive(config['data_dir'], pattern='TEST', config=config)

119it [00:03, 29.85it/s]
66it [00:01, 37.69it/s] 


In [20]:
data.feature_df

dim_0     dim_1     dim_2     dim_3     dim_4     dim_5     dim_6  \
0  -0.269570 -2.153832 -0.259319 -2.121628 -0.999057 -1.514857 -0.514935   
0  -0.269570 -2.089203 -0.288860 -2.089363 -0.999057 -1.497081 -0.514935   
0  -0.269570 -2.056059 -0.303630 -2.027255 -0.999057 -1.464492 -0.528758   
0  -0.287474 -2.006345 -0.330216 -1.962726 -0.999057 -1.464492 -0.528758   
0  -0.323280 -1.988116 -0.343509 -1.907877 -0.999057 -1.464492 -0.544116   
..       ...       ...       ...       ...       ...       ...       ...   
94 -0.104462 -1.923486 -0.166268 -1.946594 -0.907768 -1.286734 -0.408963   
94 -0.144247 -2.054402 -0.181038 -2.034514 -0.872534 -1.464492 -0.456574   
94 -0.166129 -2.130632 -0.210578 -2.113562 -0.939800 -1.545471 -0.487290   
94 -0.209893 -2.213490 -0.259319 -2.146632 -0.970229 -1.556334 -0.487290   
94 -0.229785 -2.186975 -0.259319 -2.164378 -0.999057 -1.534608 -0.487290   

       dim_7     dim_8     dim_9  ...    dim_14    dim_15    dim_16    dim_17  \
0  -1.837894 -1.126190 -0.656912  ...  0.267595 -1.654712 -0.454857 -0.464965   
0  -1.823090 -1.126190 -0.672042  ...  0.234318 -1.654712 -0.454857 -0.456310   
0  -1.786821 -1.088955 -0.668638  ...  0.199456 -1.640794 -0.454857 -0.456310   
0  -1.758695 -1.126190 -0.675447  ...  0.151917 -1.616753 -0.472082 -0.464965   
0  -1.726867 -1.126190 -0.686795  ...  0.123394 -1.628141 -0.472082 -0.464965   
..       ...       ...       ...  ...       ...       ...       ...       ...   
94 -1.883045 -1.126190 -0.545512  ...  0.437149 -1.565509 -0.454857 -0.503523   
94 -1.853438 -1.126190 -0.577475  ...  0.405457 -1.589549 -0.454857 -0.489359   
94 -1.899329 -1.088955 -0.592606  ...  0.375349 -1.602835 -0.454857 -0.482277   
94 -1.862320 -1.107573 -0.618706  ...  0.346826 -1.616753 -0.454857 -0.482277   
94 -1.862320 -1.107573 -0.641403  ...  0.299287 -1.654712 -0.454857 -0.482277   

      dim_18    dim_19    dim_20    dim_21    dim_22    dim_23  
0  -0.418045  0.330620 -1.948513  1.202557  0.106478 -1.981361  
0  -0.436693  0.342184 -1.915849  1.134967  0.106478 -1.973379  
0  -0.455342  0.353748 -1.883185  1.067377  0.106478 -1.965397  
0  -0.473990  0.365312 -1.850521  0.999787  0.106478 -1.957415  
0  -0.492638  0.376875 -1.817857  0.932197  0.106478 -1.949433  
..       ...       ...       ...       ...       ...       ...  
94 -0.371424  0.307492 -2.111834  1.540506  0.106478 -2.101089  
94 -0.380749  0.312118 -2.079170  1.472917  0.106478 -2.077144  
94 -0.390073  0.316744 -2.046506  1.405327  0.106478 -2.053198  
94 -0.399397  0.321369 -2.013842  1.337737  0.106478 -2.029252  
94 -0.408721  0.325995 -1.981178  1.270147  0.106478 -2.005306  

[13680 rows x 24 columns]

In [24]:
train_indices, val_indices, _ = split_dataset(data_indices=data.all_IDs,
                                                         validation_method='ShuffleSplit',
                                                         n_splits=1,
                                                         validation_ratio=0.2,
                                                         test_set_ratio=0,  # used only if test_indices not explicitly specified
                                                         test_indices=None,
                                                         random_seed=1337,
                                                         labels=None)
train_indices = train_indices[0]
val_indices = val_indices[0]
test_indices = np.array(test_data.all_IDs)

In [25]:
normalizer = Normalizer(config['normalization'])
data.feature_df = normalizer.normalize(data.feature_df)
test_data.feature_df = normalizer.normalize(test_data.feature_df)

In [1135]:
config['task'] = 'simmtm'
config['normalization_layer'] = 'BatchNorm'
config['out_len'] = 1
config['out_dim'] = 1
config['d_model'] = 8
config['dim_feedforward'] = 16
config['num_heads'] = 4
config['num_layers'] = 2

# SimMTMTransformerEncoder
from models.ts_transformer import model_factory
model = model_factory(config, data)
device = "cuda"
model.to(device)
model.tau = 0.05
min_tau = 0.05
model.mask_length = data.feature_df.loc[0].shape[0]//4
print(model.mask_length)
model.mask_ratio = .5

36


In [1136]:
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-3)

In [1137]:
from torch.utils.data import DataLoader
batch_size = 64
max_len = data.feature_df.loc[0].shape[0]
train_dataloader = DataLoader(train_indices, batch_size=batch_size, shuffle=True)
val_dataloader = DataLoader(val_indices, batch_size=batch_size, shuffle=True)
test_dataloader = DataLoader(test_indices, batch_size=batch_size, shuffle=True)



# Pre-Training Loop

In [1138]:

i = 0
max_epoch = 0
best_loss = 1e10
best_model = copy.deepcopy(model)
best_epoch = 0
loss_fn = nn.MSELoss()


while i < max_epoch:
    train_loss = { "loss": [], "loss_mse": [], "loss_con": []}
    progress_bar = tqdm(train_dataloader)
    
    for IDs in progress_bar:
        model.train()
        X = torch.tensor(data.feature_df.loc[IDs].to_numpy()).to(device)
        X = X.float()
        X = X.reshape(-1, max_len, X.shape[-1])
        
        pred, R, w1, w2, norm = model(X)  # (batch_size, padded_length, feat_dim)
        
        loss_mse = loss_fn(pred, X) 

        loss_con = contrastive_loss(R, X.shape[0])
        
        loss_norm =  norm.mean()

        loss = 1/(w1.pow(2)) * loss_mse + 1/(w2.pow(2)) * loss_con + torch.log(w1) + torch.log(w2) 
  


        optimizer.zero_grad()
        loss.backward()

        nn.utils.clip_grad_norm_(model.parameters(), max_norm=4.0)
        optimizer.step()

        progress_bar.set_description("Epoch {0} - Training loss: {1:.2f} - MSE loss: {2:.2f} - Contrastive loss: {3:.2f} - norm loss: {3:.2f}".format(i, 
                loss.cpu().detach().numpy().item(), loss_mse.cpu().detach().numpy().item(), 
                loss_con.cpu().detach().numpy().item(), loss_norm.cpu().detach().numpy().item())) 
        train_loss["loss"].append(loss)
        train_loss["loss_mse"].append(loss_mse)
        train_loss["loss_con"].append(loss_con)
    model.tau = max(model.tau * 0.98, min_tau)
        
    
    val_loss = { "loss": [], "loss_mse": [], "loss_con": []}
    for IDs in val_dataloader:
        model.eval()
        X = torch.tensor(data.feature_df.loc[IDs].to_numpy()).to(device)
        X = X.float()
        X = X.reshape(-1, max_len, X.shape[-1])
        
                
        pred, R, w1, w2, norm = model(X.float())  # (batch_size, padded_length, feat_dim)
        
        loss_mse = loss_fn(pred, X) 

        loss_con = contrastive_loss(R, X.shape[0])

        loss = 1/(w1.pow(2)) * loss_mse + 1/(w2.pow(2)) * loss_con + torch.log(w1) + torch.log(w2)
        
        val_loss["loss"].append(loss)
        val_loss["loss_mse"].append(loss_mse)
        val_loss["loss_con"].append(loss_con)
    
    train_loss["loss"] = torch.tensor(train_loss["loss"]).mean()
    train_loss["loss_mse"] = torch.tensor(train_loss["loss_mse"]).mean()
    train_loss["loss_con"] = torch.tensor(train_loss["loss_con"]).mean()
    val_loss["loss"] = torch.tensor(val_loss["loss"]).mean()
    val_loss["loss_mse"] = torch.tensor(val_loss["loss_mse"]).mean()
    val_loss["loss_con"] = torch.tensor(val_loss["loss_con"]).mean()
    
    if val_loss["loss"] < best_loss and model.tau == min_tau:
        best_loss = val_loss["loss"]
        best_model = copy.deepcopy(model)
        best_epoch = i
    
    progress_bar.write("Epoch {0} - Training loss: {1:.2f} {2:.2f} {3:.2f} - Validation loss: {4:.2f} {5:.2f} {6:.2f}".format(i, 
            train_loss["loss"].cpu().detach().numpy().item(), train_loss["loss_mse"].cpu().detach().numpy().item(), train_loss["loss_con"].cpu().detach().numpy().item(),
            val_loss["loss"].cpu().detach().numpy().item(), val_loss["loss_mse"].cpu().detach().numpy().item(), val_loss["loss_con"].cpu().detach().numpy().item()))
    i += 1
    
    
tqdm.write("Best Epoch {} - Best Validation loss: {}".format(best_epoch, best_loss))

Best Epoch 0 - Best Validation loss: 10000000000.0


# Replace the output layer for downstream task

In [1139]:
finetune_model = copy.deepcopy(best_model)
# finetune_model.predict_layer1 = nn.Linear(finetune_model.max_len , 1)
# out_size = data.labels_df.shape[1]
# hid_size = finetune_model.d_model 
# finetune_model.predict_layer2 = nn.Linear(hid_size , out_size)

In [1140]:


optimizer = torch.optim.AdamW(finetune_model.parameters(), lr=1e-3)

# Finetune Training Loop

In [1141]:
i = 0
max_epoch = 400
best_loss = 1e10
best_finetune_model = None
best_epoch = 0
device = "cuda"
finetune_model.to(device)
while i < max_epoch:
    train_loss = []
    progress_bar = tqdm(train_dataloader)
    
    for IDs in progress_bar:
        finetune_model.train()
        X = torch.tensor(data.feature_df.loc[IDs].to_numpy()).to(device)
        X = X.reshape(-1, max_len, X.shape[-1])
        targets = torch.tensor(data.labels_df.loc[IDs].to_numpy()).to(device)
        pred = finetune_model.predict(X.float()).squeeze(-1)
        loss = loss_fn(pred, targets)


        optimizer.zero_grad()
        loss.backward()

        nn.utils.clip_grad_norm_(finetune_model.parameters(), max_norm=4.0)
        optimizer.step()

        progress_bar.set_description("Epoch {} - Training loss: {}".format(i, loss)) 
        train_loss.append(loss)
    
    val_loss = []
    for IDs in val_dataloader:
        finetune_model.eval()
        X = torch.tensor(data.feature_df.loc[IDs].to_numpy()).to(device)
        X = X.reshape(-1, max_len, X.shape[-1])
        targets = torch.tensor(data.labels_df.loc[IDs].to_numpy()).to(device)
        pred = finetune_model.predict(X.float()).squeeze(-1)
        loss = loss_fn(pred, targets)
        val_loss.append(loss)
    
    train_loss = torch.tensor(train_loss).mean()
    val_loss = torch.tensor(val_loss).mean()
    
    if val_loss < best_loss:
        best_loss = val_loss
        best_finetune_model = copy.deepcopy(finetune_model)
        best_epoch = i
    
    progress_bar.write("Epoch {} - Training loss: {} - Validation loss: {}".format(i, train_loss, val_loss))
    i += 1
    
    
tqdm.write("Best Epoch {} - Best Validation loss: {}".format(best_epoch, best_loss))

Epoch 4 - Training loss: 183.98291015625: 100%|███████████████████████████████████████| 2/2 [00:00<00:00, 51.66it/s]

Epoch 0 - Training loss: 186.9484100341797 - Validation loss: 269.57745361328125
Epoch 1 - Training loss: 219.54864501953125 - Validation loss: 268.97491455078125
Epoch 2 - Training loss: 207.2030792236328 - Validation loss: 268.5617370605469
Epoch 3 - Training loss: 182.6446533203125 - Validation loss: 268.2022399902344



Epoch 8 - Training loss: 184.15869140625: 100%|███████████████████████████████████████| 2/2 [00:00<00:00, 51.39it/s]

Epoch 4 - Training loss: 194.3675994873047 - Validation loss: 267.8271484375
Epoch 5 - Training loss: 188.67726135253906 - Validation loss: 267.3499450683594
Epoch 6 - Training loss: 212.71441650390625 - Validation loss: 266.77508544921875
Epoch 7 - Training loss: 199.86184692382812 - Validation loss: 266.00897216796875



Epoch 12 - Training loss: 167.44949340820312: 100%|███████████████████████████████████| 2/2 [00:00<00:00, 52.87it/s]

Epoch 8 - Training loss: 192.90440368652344 - Validation loss: 265.0201721191406
Epoch 9 - Training loss: 178.9453125 - Validation loss: 263.71234130859375
Epoch 10 - Training loss: 209.99520874023438 - Validation loss: 262.2259216308594
Epoch 11 - Training loss: 191.51895141601562 - Validation loss: 260.31243896484375



Epoch 16 - Training loss: 143.93807983398438: 100%|███████████████████████████████████| 2/2 [00:00<00:00, 54.65it/s]

Epoch 12 - Training loss: 183.1023712158203 - Validation loss: 258.3831787109375
Epoch 13 - Training loss: 193.28158569335938 - Validation loss: 255.53868103027344
Epoch 14 - Training loss: 213.42108154296875 - Validation loss: 252.7337646484375
Epoch 15 - Training loss: 193.1585693359375 - Validation loss: 249.6211700439453



  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 16 - Training loss: 169.48931884765625 - Validation loss: 246.40943908691406
Epoch 17 - Training loss: 176.81326293945312 - Validation loss: 242.8572235107422
Epoch 18 - Training loss: 177.8387451171875 - Validation loss: 239.03875732421875
Epoch 19 - Training loss: 170.74839782714844 - Validation loss: 235.30709838867188
Epoch 20 - Training loss: 176.03323364257812 - Validation loss: 230.9837646484375


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 21 - Training loss: 191.33609008789062 - Validation loss: 226.02120971679688
Epoch 22 - Training loss: 178.3848114013672 - Validation loss: 221.01690673828125
Epoch 23 - Training loss: 173.7371368408203 - Validation loss: 215.99879455566406
Epoch 24 - Training loss: 156.65777587890625 - Validation loss: 211.6431121826172
Epoch 25 - Training loss: 154.71615600585938 - Validation loss: 208.03341674804688


Epoch 30 - Training loss: 130.1192169189453:   0%|                                            | 0/2 [00:00<?, ?it/s]

Epoch 26 - Training loss: 173.50259399414062 - Validation loss: 202.22799682617188
Epoch 27 - Training loss: 129.92593383789062 - Validation loss: 197.62168884277344
Epoch 28 - Training loss: 131.73497009277344 - Validation loss: 192.7593994140625
Epoch 29 - Training loss: 132.4754638671875 - Validation loss: 187.29469299316406


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 30 - Training loss: 157.33621215820312 - Validation loss: 181.60037231445312
Epoch 31 - Training loss: 139.78909301757812 - Validation loss: 176.5928192138672
Epoch 32 - Training loss: 122.56369018554688 - Validation loss: 169.49876403808594
Epoch 33 - Training loss: 132.39254760742188 - Validation loss: 161.95822143554688
Epoch 34 - Training loss: 106.37459564208984 - Validation loss: 157.71885681152344


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 35 - Training loss: 111.35604858398438 - Validation loss: 152.18524169921875
Epoch 36 - Training loss: 102.05886840820312 - Validation loss: 146.02099609375
Epoch 37 - Training loss: 113.10193634033203 - Validation loss: 141.18301391601562
Epoch 38 - Training loss: 130.55978393554688 - Validation loss: 135.8275146484375
Epoch 39 - Training loss: 107.00192260742188 - Validation loss: 131.5140838623047


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 40 - Training loss: 109.37985229492188 - Validation loss: 126.36265563964844
Epoch 41 - Training loss: 84.24246978759766 - Validation loss: 122.29254913330078
Epoch 42 - Training loss: 94.82038116455078 - Validation loss: 115.14083099365234
Epoch 43 - Training loss: 72.39570617675781 - Validation loss: 112.10725402832031
Epoch 44 - Training loss: 76.93592834472656 - Validation loss: 106.96403503417969


Epoch 49 - Training loss: 62.157554626464844: 100%|███████████████████████████████████| 2/2 [00:00<00:00, 52.65it/s]

Epoch 45 - Training loss: 72.70980072021484 - Validation loss: 102.52354431152344
Epoch 46 - Training loss: 78.46673583984375 - Validation loss: 96.92147064208984
Epoch 47 - Training loss: 70.41504669189453 - Validation loss: 91.70185089111328
Epoch 48 - Training loss: 51.45787811279297 - Validation loss: 86.17292022705078



Epoch 53 - Training loss: 59.81045150756836: 100%|████████████████████████████████████| 2/2 [00:00<00:00, 52.06it/s]

Epoch 49 - Training loss: 57.956825256347656 - Validation loss: 79.37214660644531
Epoch 50 - Training loss: 63.435062408447266 - Validation loss: 74.82440948486328
Epoch 51 - Training loss: 37.53828811645508 - Validation loss: 68.17162322998047
Epoch 52 - Training loss: 38.33171844482422 - Validation loss: 61.91176223754883



Epoch 57 - Training loss: 39.10072326660156: 100%|████████████████████████████████████| 2/2 [00:00<00:00, 51.35it/s]

Epoch 53 - Training loss: 51.67734146118164 - Validation loss: 57.414154052734375
Epoch 54 - Training loss: 33.885047912597656 - Validation loss: 53.06871032714844
Epoch 55 - Training loss: 38.249488830566406 - Validation loss: 49.2081184387207
Epoch 56 - Training loss: 35.87453079223633 - Validation loss: 46.66842269897461



Epoch 61 - Training loss: 15.98536205291748: 100%|████████████████████████████████████| 2/2 [00:00<00:00, 52.97it/s]

Epoch 57 - Training loss: 35.85618209838867 - Validation loss: 43.825889587402344
Epoch 58 - Training loss: 38.80445861816406 - Validation loss: 40.832725524902344
Epoch 59 - Training loss: 23.550487518310547 - Validation loss: 38.39033889770508
Epoch 60 - Training loss: 35.99459457397461 - Validation loss: 36.63348388671875



Epoch 65 - Training loss: 18.203407287597656: 100%|███████████████████████████████████| 2/2 [00:00<00:00, 53.24it/s]

Epoch 61 - Training loss: 22.635469436645508 - Validation loss: 35.30989074707031
Epoch 62 - Training loss: 32.18497848510742 - Validation loss: 35.2609748840332
Epoch 63 - Training loss: 23.07387924194336 - Validation loss: 33.862945556640625
Epoch 64 - Training loss: 20.088848114013672 - Validation loss: 32.610599517822266



  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 65 - Training loss: 19.778156280517578 - Validation loss: 32.48920822143555
Epoch 66 - Training loss: 26.036949157714844 - Validation loss: 33.09881591796875
Epoch 67 - Training loss: 21.698482513427734 - Validation loss: 33.5000114440918
Epoch 68 - Training loss: 20.277332305908203 - Validation loss: 33.8072624206543
Epoch 69 - Training loss: 16.788663864135742 - Validation loss: 33.79048156738281


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 70 - Training loss: 18.898513793945312 - Validation loss: 33.422977447509766
Epoch 71 - Training loss: 22.35212516784668 - Validation loss: 33.02914810180664
Epoch 72 - Training loss: 22.224275588989258 - Validation loss: 31.736759185791016
Epoch 73 - Training loss: 22.938209533691406 - Validation loss: 31.079082489013672
Epoch 74 - Training loss: 26.609739303588867 - Validation loss: 30.765005111694336


Epoch 80 - Training loss: 17.88681983947754:   0%|                                            | 0/2 [00:00<?, ?it/s]

Epoch 75 - Training loss: 23.30147361755371 - Validation loss: 28.691091537475586
Epoch 76 - Training loss: 24.31298828125 - Validation loss: 26.957618713378906
Epoch 77 - Training loss: 27.57733154296875 - Validation loss: 27.9455623626709
Epoch 78 - Training loss: 21.544506072998047 - Validation loss: 28.660207748413086
Epoch 79 - Training loss: 15.372241973876953 - Validation loss: 29.775346755981445


Epoch 85 - Training loss: 19.320663452148438:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 80 - Training loss: 13.427541732788086 - Validation loss: 30.129854202270508
Epoch 81 - Training loss: 20.32022476196289 - Validation loss: 31.561405181884766
Epoch 82 - Training loss: 15.510414123535156 - Validation loss: 31.964946746826172
Epoch 83 - Training loss: 22.98802947998047 - Validation loss: 30.90745735168457
Epoch 84 - Training loss: 15.906929016113281 - Validation loss: 29.998531341552734


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 85 - Training loss: 15.472127914428711 - Validation loss: 28.949296951293945
Epoch 86 - Training loss: 18.024181365966797 - Validation loss: 27.7235107421875
Epoch 87 - Training loss: 21.175777435302734 - Validation loss: 26.5753116607666
Epoch 88 - Training loss: 15.674163818359375 - Validation loss: 25.037517547607422
Epoch 89 - Training loss: 14.963090896606445 - Validation loss: 23.20470428466797


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 90 - Training loss: 21.490718841552734 - Validation loss: 22.29348373413086
Epoch 91 - Training loss: 11.335250854492188 - Validation loss: 22.186458587646484
Epoch 92 - Training loss: 19.581117630004883 - Validation loss: 22.675424575805664
Epoch 93 - Training loss: 17.03508758544922 - Validation loss: 24.043432235717773
Epoch 94 - Training loss: 12.695184707641602 - Validation loss: 25.193790435791016


Epoch 100 - Training loss: 21.531375885009766:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 95 - Training loss: 18.265178680419922 - Validation loss: 26.423364639282227
Epoch 96 - Training loss: 16.988327026367188 - Validation loss: 28.41969108581543
Epoch 97 - Training loss: 16.74444580078125 - Validation loss: 28.984542846679688
Epoch 98 - Training loss: 22.25795555114746 - Validation loss: 27.308330535888672
Epoch 99 - Training loss: 15.164043426513672 - Validation loss: 24.597272872924805


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 100 - Training loss: 17.49856185913086 - Validation loss: 21.908037185668945
Epoch 101 - Training loss: 13.383262634277344 - Validation loss: 19.549034118652344
Epoch 102 - Training loss: 19.586328506469727 - Validation loss: 18.492820739746094
Epoch 103 - Training loss: 13.908712387084961 - Validation loss: 18.064001083374023
Epoch 104 - Training loss: 13.96010684967041 - Validation loss: 18.83076286315918


Epoch 110 - Training loss: 12.094075202941895:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 105 - Training loss: 13.305716514587402 - Validation loss: 19.656667709350586
Epoch 106 - Training loss: 25.963394165039062 - Validation loss: 20.483938217163086
Epoch 107 - Training loss: 14.113056182861328 - Validation loss: 20.448823928833008
Epoch 108 - Training loss: 12.28256893157959 - Validation loss: 21.136178970336914
Epoch 109 - Training loss: 17.029216766357422 - Validation loss: 22.033096313476562


Epoch 115 - Training loss: 16.7357177734375:   0%|                                            | 0/2 [00:00<?, ?it/s]

Epoch 110 - Training loss: 11.803194046020508 - Validation loss: 21.303585052490234
Epoch 111 - Training loss: 18.579912185668945 - Validation loss: 21.097225189208984
Epoch 112 - Training loss: 15.787991523742676 - Validation loss: 21.207494735717773
Epoch 113 - Training loss: 12.915653228759766 - Validation loss: 21.01564598083496
Epoch 114 - Training loss: 14.30449390411377 - Validation loss: 20.622594833374023


Epoch 120 - Training loss: 15.652176856994629:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 115 - Training loss: 16.4990234375 - Validation loss: 20.06017303466797
Epoch 116 - Training loss: 12.926928520202637 - Validation loss: 20.01369285583496
Epoch 117 - Training loss: 19.069828033447266 - Validation loss: 20.302616119384766
Epoch 118 - Training loss: 21.902912139892578 - Validation loss: 20.922765731811523
Epoch 119 - Training loss: 18.310636520385742 - Validation loss: 20.857336044311523


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 120 - Training loss: 14.821712493896484 - Validation loss: 21.660062789916992
Epoch 121 - Training loss: 19.466379165649414 - Validation loss: 22.538183212280273
Epoch 122 - Training loss: 11.491591453552246 - Validation loss: 22.876567840576172
Epoch 123 - Training loss: 14.345450401306152 - Validation loss: 22.28602409362793
Epoch 124 - Training loss: 15.400781631469727 - Validation loss: 21.58322525024414


Epoch 130 - Training loss: 15.558248519897461:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 125 - Training loss: 11.343017578125 - Validation loss: 21.569503784179688
Epoch 126 - Training loss: 20.314781188964844 - Validation loss: 20.960458755493164
Epoch 127 - Training loss: 9.33187484741211 - Validation loss: 21.038257598876953
Epoch 128 - Training loss: 11.672727584838867 - Validation loss: 21.579938888549805
Epoch 129 - Training loss: 18.862815856933594 - Validation loss: 22.440744400024414


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 130 - Training loss: 13.24697494506836 - Validation loss: 23.18331527709961
Epoch 131 - Training loss: 14.987098693847656 - Validation loss: 24.270824432373047
Epoch 132 - Training loss: 9.603650093078613 - Validation loss: 24.262481689453125
Epoch 133 - Training loss: 10.976604461669922 - Validation loss: 23.471336364746094
Epoch 134 - Training loss: 17.96708106994629 - Validation loss: 23.219472885131836


Epoch 140 - Training loss: 13.35884952545166:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 135 - Training loss: 12.393696784973145 - Validation loss: 22.477558135986328
Epoch 136 - Training loss: 12.982887268066406 - Validation loss: 21.24087905883789
Epoch 137 - Training loss: 12.529274940490723 - Validation loss: 20.142005920410156
Epoch 138 - Training loss: 16.77733039855957 - Validation loss: 19.999486923217773
Epoch 139 - Training loss: 15.918484687805176 - Validation loss: 20.182052612304688


Epoch 145 - Training loss: 17.479568481445312:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 140 - Training loss: 11.822771072387695 - Validation loss: 20.58588218688965
Epoch 141 - Training loss: 12.870416641235352 - Validation loss: 20.875741958618164
Epoch 142 - Training loss: 21.16647720336914 - Validation loss: 20.774967193603516
Epoch 143 - Training loss: 10.66209602355957 - Validation loss: 21.246234893798828
Epoch 144 - Training loss: 14.765057563781738 - Validation loss: 20.96959114074707


Epoch 150 - Training loss: 10.598352432250977:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 145 - Training loss: 15.14859390258789 - Validation loss: 20.416128158569336
Epoch 146 - Training loss: 13.360764503479004 - Validation loss: 20.869665145874023
Epoch 147 - Training loss: 14.307929992675781 - Validation loss: 20.264080047607422
Epoch 148 - Training loss: 13.593425750732422 - Validation loss: 19.300817489624023
Epoch 149 - Training loss: 14.805073738098145 - Validation loss: 18.88707160949707


Epoch 155 - Training loss: 12.872557640075684:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 150 - Training loss: 13.554583549499512 - Validation loss: 18.717269897460938
Epoch 151 - Training loss: 14.227897644042969 - Validation loss: 18.510744094848633
Epoch 152 - Training loss: 12.158666610717773 - Validation loss: 18.79706382751465
Epoch 153 - Training loss: 9.532302856445312 - Validation loss: 18.96121597290039
Epoch 154 - Training loss: 12.855730056762695 - Validation loss: 18.520971298217773


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 155 - Training loss: 13.377501487731934 - Validation loss: 19.106948852539062
Epoch 156 - Training loss: 12.31491756439209 - Validation loss: 19.392837524414062
Epoch 157 - Training loss: 12.648871421813965 - Validation loss: 18.3752498626709
Epoch 158 - Training loss: 11.488266944885254 - Validation loss: 18.093807220458984
Epoch 159 - Training loss: 11.800566673278809 - Validation loss: 17.529542922973633


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 160 - Training loss: 11.037668228149414 - Validation loss: 17.625661849975586
Epoch 161 - Training loss: 9.659984588623047 - Validation loss: 17.889141082763672
Epoch 162 - Training loss: 10.682855606079102 - Validation loss: 17.87017059326172
Epoch 163 - Training loss: 11.793649673461914 - Validation loss: 18.147518157958984
Epoch 164 - Training loss: 10.132364273071289 - Validation loss: 18.53609275817871


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 165 - Training loss: 14.857695579528809 - Validation loss: 18.37118148803711
Epoch 166 - Training loss: 13.307369232177734 - Validation loss: 18.29216957092285
Epoch 167 - Training loss: 8.842083930969238 - Validation loss: 18.469453811645508
Epoch 168 - Training loss: 15.369768142700195 - Validation loss: 18.502761840820312
Epoch 169 - Training loss: 12.382576942443848 - Validation loss: 19.15439224243164


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 170 - Training loss: 11.209001541137695 - Validation loss: 18.62995147705078
Epoch 171 - Training loss: 15.981328010559082 - Validation loss: 18.3636474609375
Epoch 172 - Training loss: 11.293859481811523 - Validation loss: 17.875905990600586
Epoch 173 - Training loss: 14.66091251373291 - Validation loss: 17.442401885986328
Epoch 174 - Training loss: 9.354324340820312 - Validation loss: 15.962930679321289


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 175 - Training loss: 13.807085990905762 - Validation loss: 14.99425983428955
Epoch 176 - Training loss: 10.689990997314453 - Validation loss: 14.827237129211426
Epoch 177 - Training loss: 13.593182563781738 - Validation loss: 14.74407958984375
Epoch 178 - Training loss: 14.55361270904541 - Validation loss: 14.745161056518555
Epoch 179 - Training loss: 22.578811645507812 - Validation loss: 15.242640495300293


Epoch 185 - Training loss: 12.231487274169922:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 180 - Training loss: 15.495285034179688 - Validation loss: 16.495864868164062
Epoch 181 - Training loss: 14.269417762756348 - Validation loss: 17.868728637695312
Epoch 182 - Training loss: 13.996790885925293 - Validation loss: 19.668415069580078
Epoch 183 - Training loss: 9.83513069152832 - Validation loss: 20.4976749420166
Epoch 184 - Training loss: 10.498701095581055 - Validation loss: 21.219179153442383


Epoch 190 - Training loss: 13.424647331237793:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 185 - Training loss: 13.360270500183105 - Validation loss: 21.655004501342773
Epoch 186 - Training loss: 17.61558723449707 - Validation loss: 21.495107650756836
Epoch 187 - Training loss: 8.784446716308594 - Validation loss: 20.199886322021484
Epoch 188 - Training loss: 19.88098907470703 - Validation loss: 19.144699096679688
Epoch 189 - Training loss: 14.039400100708008 - Validation loss: 18.18797492980957


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 190 - Training loss: 11.77371597290039 - Validation loss: 17.617122650146484
Epoch 191 - Training loss: 12.66621208190918 - Validation loss: 17.23444366455078
Epoch 192 - Training loss: 14.51264762878418 - Validation loss: 16.812761306762695
Epoch 193 - Training loss: 11.796597480773926 - Validation loss: 16.679508209228516
Epoch 194 - Training loss: 16.047712326049805 - Validation loss: 16.626951217651367


Epoch 200 - Training loss: 13.855205535888672:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 195 - Training loss: 9.864885330200195 - Validation loss: 16.621450424194336
Epoch 196 - Training loss: 9.4114990234375 - Validation loss: 16.524030685424805
Epoch 197 - Training loss: 9.584651947021484 - Validation loss: 16.917034149169922
Epoch 198 - Training loss: 9.546954154968262 - Validation loss: 17.46876335144043
Epoch 199 - Training loss: 15.464400291442871 - Validation loss: 19.0067195892334


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 200 - Training loss: 10.97648811340332 - Validation loss: 19.160863876342773
Epoch 201 - Training loss: 13.72439956665039 - Validation loss: 19.211936950683594
Epoch 202 - Training loss: 25.96700668334961 - Validation loss: 18.6109561920166
Epoch 203 - Training loss: 17.337642669677734 - Validation loss: 17.63429069519043
Epoch 204 - Training loss: 12.184661865234375 - Validation loss: 17.673816680908203


Epoch 210 - Training loss: 10.596879959106445:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 205 - Training loss: 12.279666900634766 - Validation loss: 18.17154884338379
Epoch 206 - Training loss: 8.38862419128418 - Validation loss: 18.702943801879883
Epoch 207 - Training loss: 12.86073112487793 - Validation loss: 19.390865325927734
Epoch 208 - Training loss: 8.6165189743042 - Validation loss: 19.76180076599121
Epoch 209 - Training loss: 14.113495826721191 - Validation loss: 20.02134895324707


Epoch 215 - Training loss: 9.194246292114258:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 210 - Training loss: 16.142444610595703 - Validation loss: 20.38372802734375
Epoch 211 - Training loss: 11.203102111816406 - Validation loss: 20.10726547241211
Epoch 212 - Training loss: 10.168227195739746 - Validation loss: 19.02408218383789
Epoch 213 - Training loss: 8.147683143615723 - Validation loss: 18.075748443603516
Epoch 214 - Training loss: 21.770790100097656 - Validation loss: 17.776464462280273


Epoch 220 - Training loss: 8.649560928344727:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 215 - Training loss: 8.75617790222168 - Validation loss: 17.569101333618164
Epoch 216 - Training loss: 10.464424133300781 - Validation loss: 17.427461624145508
Epoch 217 - Training loss: 17.9036865234375 - Validation loss: 17.632478713989258
Epoch 218 - Training loss: 9.384771347045898 - Validation loss: 18.21049690246582
Epoch 219 - Training loss: 7.3363542556762695 - Validation loss: 19.23196029663086


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 220 - Training loss: 12.64668083190918 - Validation loss: 20.102025985717773
Epoch 221 - Training loss: 20.755924224853516 - Validation loss: 21.43985939025879
Epoch 222 - Training loss: 7.79200553894043 - Validation loss: 21.748260498046875
Epoch 223 - Training loss: 9.454575538635254 - Validation loss: 21.445919036865234
Epoch 224 - Training loss: 9.617263793945312 - Validation loss: 20.2069034576416


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 225 - Training loss: 10.707635879516602 - Validation loss: 19.053119659423828
Epoch 226 - Training loss: 15.953800201416016 - Validation loss: 18.497228622436523
Epoch 227 - Training loss: 11.009529113769531 - Validation loss: 18.24053382873535
Epoch 228 - Training loss: 13.629613876342773 - Validation loss: 18.577110290527344
Epoch 229 - Training loss: 9.835587501525879 - Validation loss: 18.664649963378906


Epoch 235 - Training loss: 15.587625503540039:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 230 - Training loss: 10.764875411987305 - Validation loss: 18.480833053588867
Epoch 231 - Training loss: 12.575965881347656 - Validation loss: 18.30156707763672
Epoch 232 - Training loss: 10.548125267028809 - Validation loss: 18.09501838684082
Epoch 233 - Training loss: 12.1585054397583 - Validation loss: 18.161191940307617
Epoch 234 - Training loss: 10.210338592529297 - Validation loss: 18.739009857177734


Epoch 240 - Training loss: 12.343155860900879:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 235 - Training loss: 12.854488372802734 - Validation loss: 19.16664695739746
Epoch 236 - Training loss: 14.797867774963379 - Validation loss: 19.34747886657715
Epoch 237 - Training loss: 11.573406219482422 - Validation loss: 18.33256721496582
Epoch 238 - Training loss: 11.621809005737305 - Validation loss: 17.62974739074707
Epoch 239 - Training loss: 8.68091106414795 - Validation loss: 17.439340591430664


Epoch 245 - Training loss: 11.354705810546875:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 240 - Training loss: 12.455390930175781 - Validation loss: 17.37862777709961
Epoch 241 - Training loss: 15.99663257598877 - Validation loss: 17.207841873168945
Epoch 242 - Training loss: 9.88789176940918 - Validation loss: 17.733884811401367
Epoch 243 - Training loss: 13.440790176391602 - Validation loss: 18.447004318237305
Epoch 244 - Training loss: 13.796953201293945 - Validation loss: 18.71164321899414


Epoch 250 - Training loss: 14.814705848693848:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 245 - Training loss: 9.159950256347656 - Validation loss: 18.757631301879883
Epoch 246 - Training loss: 10.242111206054688 - Validation loss: 18.578475952148438
Epoch 247 - Training loss: 14.038530349731445 - Validation loss: 18.978107452392578
Epoch 248 - Training loss: 9.529119491577148 - Validation loss: 19.221174240112305
Epoch 249 - Training loss: 10.856910705566406 - Validation loss: 19.79222297668457


Epoch 255 - Training loss: 17.48625373840332:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 250 - Training loss: 11.642349243164062 - Validation loss: 20.071693420410156
Epoch 251 - Training loss: 10.098301887512207 - Validation loss: 20.648576736450195
Epoch 252 - Training loss: 8.203899383544922 - Validation loss: 20.81451988220215
Epoch 253 - Training loss: 10.645853042602539 - Validation loss: 19.976051330566406
Epoch 254 - Training loss: 9.764785766601562 - Validation loss: 18.887361526489258


Epoch 260 - Training loss: 7.747417449951172:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 255 - Training loss: 17.158599853515625 - Validation loss: 18.07257080078125
Epoch 256 - Training loss: 9.22412109375 - Validation loss: 17.425567626953125
Epoch 257 - Training loss: 9.371143341064453 - Validation loss: 17.315343856811523
Epoch 258 - Training loss: 10.604923248291016 - Validation loss: 17.37040901184082
Epoch 259 - Training loss: 8.907837867736816 - Validation loss: 17.7952938079834


Epoch 265 - Training loss: 9.134440422058105:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 260 - Training loss: 14.962167739868164 - Validation loss: 18.719497680664062
Epoch 261 - Training loss: 11.419978141784668 - Validation loss: 20.16931915283203
Epoch 262 - Training loss: 7.7433295249938965 - Validation loss: 20.467769622802734
Epoch 263 - Training loss: 10.437847137451172 - Validation loss: 21.152286529541016
Epoch 264 - Training loss: 19.697776794433594 - Validation loss: 22.154294967651367


Epoch 270 - Training loss: 11.654277801513672:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 265 - Training loss: 12.347224235534668 - Validation loss: 21.82781982421875
Epoch 266 - Training loss: 14.308956146240234 - Validation loss: 21.99224090576172
Epoch 267 - Training loss: 13.753994941711426 - Validation loss: 21.392696380615234
Epoch 268 - Training loss: 15.106889724731445 - Validation loss: 21.044090270996094
Epoch 269 - Training loss: 13.905112266540527 - Validation loss: 20.214292526245117


Epoch 274 - Training loss: 10.592086791992188: 100%|██████████████████████████████████| 2/2 [00:00<00:00, 54.67it/s]

Epoch 270 - Training loss: 10.992504119873047 - Validation loss: 19.72759437561035
Epoch 271 - Training loss: 19.997413635253906 - Validation loss: 19.078166961669922
Epoch 272 - Training loss: 14.873795509338379 - Validation loss: 18.966218948364258
Epoch 273 - Training loss: 13.351114273071289 - Validation loss: 18.5823917388916



Epoch 278 - Training loss: 6.699913024902344: 100%|███████████████████████████████████| 2/2 [00:00<00:00, 50.53it/s]

Epoch 274 - Training loss: 10.184209823608398 - Validation loss: 18.153291702270508
Epoch 275 - Training loss: 15.50571346282959 - Validation loss: 17.805465698242188
Epoch 276 - Training loss: 14.562530517578125 - Validation loss: 18.773815155029297
Epoch 277 - Training loss: 14.459819793701172 - Validation loss: 18.922780990600586



  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 278 - Training loss: 7.67074728012085 - Validation loss: 18.97933006286621
Epoch 279 - Training loss: 15.150772094726562 - Validation loss: 19.47173309326172
Epoch 280 - Training loss: 11.226926803588867 - Validation loss: 19.616209030151367
Epoch 281 - Training loss: 31.696765899658203 - Validation loss: 20.051790237426758
Epoch 282 - Training loss: 12.263792037963867 - Validation loss: 19.941640853881836


Epoch 288 - Training loss: 12.951120376586914:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 283 - Training loss: 7.540073871612549 - Validation loss: 19.73895835876465
Epoch 284 - Training loss: 12.077241897583008 - Validation loss: 18.840944290161133
Epoch 285 - Training loss: 8.66037654876709 - Validation loss: 18.4444522857666
Epoch 286 - Training loss: 9.267866134643555 - Validation loss: 18.2561092376709
Epoch 287 - Training loss: 11.595861434936523 - Validation loss: 18.912490844726562


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 288 - Training loss: 9.558232307434082 - Validation loss: 20.50057601928711
Epoch 289 - Training loss: 8.762532234191895 - Validation loss: 21.283300399780273
Epoch 290 - Training loss: 8.925163269042969 - Validation loss: 22.59425163269043
Epoch 291 - Training loss: 11.244571685791016 - Validation loss: 23.28523063659668
Epoch 292 - Training loss: 8.776069641113281 - Validation loss: 22.8410587310791


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 293 - Training loss: 10.634519577026367 - Validation loss: 19.978065490722656
Epoch 294 - Training loss: 12.922327995300293 - Validation loss: 18.471891403198242
Epoch 295 - Training loss: 14.435568809509277 - Validation loss: 18.210939407348633
Epoch 296 - Training loss: 9.866039276123047 - Validation loss: 17.716354370117188
Epoch 297 - Training loss: 11.70458984375 - Validation loss: 17.11551284790039


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 298 - Training loss: 16.189605712890625 - Validation loss: 17.753047943115234
Epoch 299 - Training loss: 15.375696182250977 - Validation loss: 18.573331832885742
Epoch 300 - Training loss: 11.746003150939941 - Validation loss: 18.795124053955078
Epoch 301 - Training loss: 9.027143478393555 - Validation loss: 18.606672286987305
Epoch 302 - Training loss: 11.708736419677734 - Validation loss: 18.169815063476562


Epoch 308 - Training loss: 14.020702362060547:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 303 - Training loss: 10.573030471801758 - Validation loss: 18.00598907470703
Epoch 304 - Training loss: 14.190996170043945 - Validation loss: 18.281883239746094
Epoch 305 - Training loss: 12.582719802856445 - Validation loss: 19.315038681030273
Epoch 306 - Training loss: 9.180777549743652 - Validation loss: 19.534622192382812
Epoch 307 - Training loss: 13.438159942626953 - Validation loss: 19.83973503112793


Epoch 313 - Training loss: 11.595515251159668:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 308 - Training loss: 11.407792091369629 - Validation loss: 19.783170700073242
Epoch 309 - Training loss: 9.069446563720703 - Validation loss: 19.75448226928711
Epoch 310 - Training loss: 8.629624366760254 - Validation loss: 20.507543563842773
Epoch 311 - Training loss: 12.92248821258545 - Validation loss: 20.671632766723633
Epoch 312 - Training loss: 9.673988342285156 - Validation loss: 20.940771102905273


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 313 - Training loss: 10.713172912597656 - Validation loss: 20.402462005615234
Epoch 314 - Training loss: 9.022933006286621 - Validation loss: 19.997047424316406
Epoch 315 - Training loss: 11.541820526123047 - Validation loss: 18.934980392456055
Epoch 316 - Training loss: 12.57675552368164 - Validation loss: 17.938846588134766
Epoch 317 - Training loss: 9.767732620239258 - Validation loss: 17.667936325073242


Epoch 323 - Training loss: 11.560836791992188:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 318 - Training loss: 9.756063461303711 - Validation loss: 17.180438995361328
Epoch 319 - Training loss: 24.742538452148438 - Validation loss: 17.49871826171875
Epoch 320 - Training loss: 18.84222412109375 - Validation loss: 17.265682220458984
Epoch 321 - Training loss: 12.221866607666016 - Validation loss: 16.65435791015625
Epoch 322 - Training loss: 16.559288024902344 - Validation loss: 16.842098236083984


Epoch 328 - Training loss: 9.587940216064453:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 323 - Training loss: 14.476411819458008 - Validation loss: 16.972970962524414
Epoch 324 - Training loss: 13.263551712036133 - Validation loss: 17.296676635742188
Epoch 325 - Training loss: 11.073934555053711 - Validation loss: 17.62543296813965
Epoch 326 - Training loss: 9.023119926452637 - Validation loss: 18.280197143554688
Epoch 327 - Training loss: 12.926216125488281 - Validation loss: 19.63033676147461


Epoch 333 - Training loss: 12.36036491394043:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 328 - Training loss: 10.135028839111328 - Validation loss: 20.936019897460938
Epoch 329 - Training loss: 19.59304428100586 - Validation loss: 21.883445739746094
Epoch 330 - Training loss: 13.300848960876465 - Validation loss: 21.85107421875
Epoch 331 - Training loss: 10.828872680664062 - Validation loss: 20.90707015991211
Epoch 332 - Training loss: 9.560708045959473 - Validation loss: 20.412572860717773


Epoch 338 - Training loss: 15.549291610717773:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 333 - Training loss: 7.828609466552734 - Validation loss: 19.771520614624023
Epoch 334 - Training loss: 14.630184173583984 - Validation loss: 18.899452209472656
Epoch 335 - Training loss: 9.990354537963867 - Validation loss: 18.011947631835938
Epoch 336 - Training loss: 10.074699401855469 - Validation loss: 17.266220092773438
Epoch 337 - Training loss: 10.814631462097168 - Validation loss: 17.14898681640625


Epoch 343 - Training loss: 13.619277000427246:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 338 - Training loss: 10.074005126953125 - Validation loss: 17.02252769470215
Epoch 339 - Training loss: 13.027769088745117 - Validation loss: 17.429025650024414
Epoch 340 - Training loss: 7.563620567321777 - Validation loss: 17.20063591003418
Epoch 341 - Training loss: 10.86694049835205 - Validation loss: 16.7418270111084
Epoch 342 - Training loss: 10.329584121704102 - Validation loss: 16.431442260742188


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 343 - Training loss: 14.178651809692383 - Validation loss: 16.70112419128418
Epoch 344 - Training loss: 14.687708854675293 - Validation loss: 17.451271057128906
Epoch 345 - Training loss: 10.083099365234375 - Validation loss: 18.488155364990234
Epoch 346 - Training loss: 8.56838321685791 - Validation loss: 19.4451961517334
Epoch 347 - Training loss: 13.003698348999023 - Validation loss: 20.660947799682617


Epoch 353 - Training loss: 10.583662033081055:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 348 - Training loss: 11.58908748626709 - Validation loss: 20.374860763549805
Epoch 349 - Training loss: 9.35310173034668 - Validation loss: 20.740663528442383
Epoch 350 - Training loss: 9.818902015686035 - Validation loss: 20.488338470458984
Epoch 351 - Training loss: 13.135039329528809 - Validation loss: 19.867198944091797
Epoch 352 - Training loss: 15.237205505371094 - Validation loss: 19.01528549194336


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 353 - Training loss: 9.2240629196167 - Validation loss: 18.02140998840332
Epoch 354 - Training loss: 9.969171524047852 - Validation loss: 17.792320251464844
Epoch 355 - Training loss: 10.09051513671875 - Validation loss: 17.745872497558594
Epoch 356 - Training loss: 21.411256790161133 - Validation loss: 17.469398498535156
Epoch 357 - Training loss: 9.296612739562988 - Validation loss: 17.53803062438965


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 358 - Training loss: 10.270454406738281 - Validation loss: 18.2546443939209
Epoch 359 - Training loss: 9.94487476348877 - Validation loss: 18.898929595947266
Epoch 360 - Training loss: 10.296119689941406 - Validation loss: 18.827987670898438
Epoch 361 - Training loss: 16.332290649414062 - Validation loss: 18.919771194458008
Epoch 362 - Training loss: 8.767496109008789 - Validation loss: 18.297927856445312


Epoch 368 - Training loss: 9.265830993652344:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 363 - Training loss: 6.79823112487793 - Validation loss: 17.537372589111328
Epoch 364 - Training loss: 10.24026107788086 - Validation loss: 17.126005172729492
Epoch 365 - Training loss: 13.4271879196167 - Validation loss: 17.587989807128906
Epoch 366 - Training loss: 19.99580955505371 - Validation loss: 17.518877029418945
Epoch 367 - Training loss: 13.855348587036133 - Validation loss: 16.88455581665039


Epoch 373 - Training loss: 7.989242076873779:   0%|                                           | 0/2 [00:00<?, ?it/s]

Epoch 368 - Training loss: 13.162609100341797 - Validation loss: 16.93230628967285
Epoch 369 - Training loss: 13.276615142822266 - Validation loss: 17.31968879699707
Epoch 370 - Training loss: 13.728260040283203 - Validation loss: 17.662235260009766
Epoch 371 - Training loss: 11.904616355895996 - Validation loss: 18.37993812561035
Epoch 372 - Training loss: 11.84097957611084 - Validation loss: 18.449861526489258


Epoch 378 - Training loss: 13.361983299255371:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 373 - Training loss: 9.464619636535645 - Validation loss: 18.135202407836914
Epoch 374 - Training loss: 11.133347511291504 - Validation loss: 17.892742156982422
Epoch 375 - Training loss: 7.338693141937256 - Validation loss: 17.202001571655273
Epoch 376 - Training loss: 22.775375366210938 - Validation loss: 17.344614028930664
Epoch 377 - Training loss: 10.127641677856445 - Validation loss: 17.647172927856445


Epoch 383 - Training loss: 12.619309425354004:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 378 - Training loss: 11.488590240478516 - Validation loss: 17.158138275146484
Epoch 379 - Training loss: 9.72359848022461 - Validation loss: 16.878942489624023
Epoch 380 - Training loss: 9.121198654174805 - Validation loss: 16.408979415893555
Epoch 381 - Training loss: 17.652278900146484 - Validation loss: 15.802018165588379
Epoch 382 - Training loss: 12.933611869812012 - Validation loss: 15.648691177368164


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 383 - Training loss: 16.64661407470703 - Validation loss: 15.446208953857422
Epoch 384 - Training loss: 9.037851333618164 - Validation loss: 15.1284818649292
Epoch 385 - Training loss: 17.244686126708984 - Validation loss: 15.001842498779297
Epoch 386 - Training loss: 10.385912895202637 - Validation loss: 15.171070098876953
Epoch 387 - Training loss: 12.119890213012695 - Validation loss: 15.705476760864258


Epoch 393 - Training loss: 10.772300720214844:   0%|                                          | 0/2 [00:00<?, ?it/s]

Epoch 388 - Training loss: 12.12100601196289 - Validation loss: 16.34530258178711
Epoch 389 - Training loss: 13.381879806518555 - Validation loss: 16.978267669677734
Epoch 390 - Training loss: 8.723334312438965 - Validation loss: 18.012990951538086
Epoch 391 - Training loss: 10.646793365478516 - Validation loss: 18.47659683227539
Epoch 392 - Training loss: 8.700498580932617 - Validation loss: 18.167831420898438


  0%|                                                                                         | 0/2 [00:00<?, ?it/s]

Epoch 393 - Training loss: 8.662198066711426 - Validation loss: 18.268199920654297
Epoch 394 - Training loss: 12.334908485412598 - Validation loss: 18.231468200683594
Epoch 395 - Training loss: 12.585148811340332 - Validation loss: 17.7148494720459
Epoch 396 - Training loss: 8.049286842346191 - Validation loss: 17.236398696899414
Epoch 397 - Training loss: 15.363054275512695 - Validation loss: 16.445581436157227


Epoch 399 - Training loss: 11.325026512145996: 100%|██████████████████████████████████| 2/2 [00:00<00:00, 47.94it/s]

Epoch 398 - Training loss: 10.398223876953125 - Validation loss: 16.20722198486328
Epoch 399 - Training loss: 12.503231048583984 - Validation loss: 15.918633460998535
Best Epoch 177 - Best Validation loss: 14.74407958984375


In [1142]:
test_loss = []
for IDs in test_dataloader:
    best_finetune_model.eval()
    X = torch.tensor(test_data.feature_df.loc[IDs].to_numpy()).to(device)
    X = X.reshape(-1, max_len, X.shape[-1])
    targets = torch.tensor(test_data.labels_df.loc[IDs].to_numpy()).to(device)
    pred = best_finetune_model.predict(X.float()).squeeze(-1)
    loss = loss_fn(pred, targets)
    test_loss.append(loss)


test_loss = torch.tensor(test_loss).mean()
print("Test loss: {}".format(test_loss))
print("Test loss: {}".format(np.sqrt(test_loss)))

Test loss: 20.3616943359375
Test loss: 4.512393474578857


# Forecasting Results from the paper

<center><img src="../img/img_7.PNG"/><center/>

Reference:
1. https://arxiv.org/abs/2302.00861
2. https://github.com/gzerveas/mvts_transformer

No Pretrain

Test loss: 17.563016891479492

Test loss: 24.951457977294922

Test loss: 13.356301307678223

Test loss: 15.543647766113281

Test loss: 10.70577621459961

Test loss: 11.653989791870117

Test loss: 28.348876953125

Test loss: 9.66756534576416

Test loss: 22.540130615234375

Test loss: 20.3616943359375

Pretrain L//2 pretrain epoch 50

Test loss: 11.152569770812988

Test loss: 8.513360023498535

Test loss: 20.0480899810791

Test loss: 12.694931983947754

Test loss: 15.458633422851562

Test loss: 16.706832885742188

Test loss: 11.432455062866211

Test loss: 7.263433456420898

Test loss: 15.908415794372559

Test loss: 17.736656188964844

Pretrain L//2 pretrain epoch 20

Test loss: 13.188956260681152

Test loss: 12.610322952270508

Test loss: 22.47460174560547

Test loss: 17.15345001220703

Test loss: 20.823780059814453

Pretrain L//2 pretrain epoch 100

Test loss: 20.534494400024414

Test loss: 25.885923385620117

Test loss: 21.357059478759766

Test loss: 20.61741828918457

Test loss: 24.453622817993164

Pretrain L//4 pretrain epoch 50

Test loss: 8.65383243560791

Test loss: 7.952672481536865

Test loss: 13.125472068786621

Test loss: 17.975181579589844

Test loss: 21.050844192504883

Test loss: 9.025917053222656

Test loss: 13.526683807373047

Test loss: 16.005821228027344

Test loss: 23.140092849731445

Test loss: 11.915889739990234

Pretrain L//10 pretrain epoch 50

Test loss: 18.413068771362305

Test loss: 17.07589340209961

Test loss: 14.856538772583008

Test loss: 9.033221244812012

Test loss: 16.976844787597656